src/mame/drivers/firebeat.cpp

```diff
@@ -176,687 +177,962 @@ struct IBUTTON
-	// 0x00020000 Stage LED 5
-	// 0x00040000 Stage LED 6
-	// 0x00080000 Stage LED 7
-	if (ACCESSING_BITS_8_15)
-	{
-		output().set_value("left",            (data & 0x00000100) ? 1 : 0);
-		output().set_value("right",           (data & 0x00000200) ? 1 : 0);
-		output().set_value("door_lamp",       (data & 0x00000400) ? 1 : 0);
-		output().set_value("ok",              (data & 0x00000800) ? 1 : 0);
-		output().set_value("slim",            (data & 0x00008000) ? 1 : 0);
-	}
-	if (ACCESSING_BITS_24_31)
-	{
-		output().set_value("stage_led_0",     (data & 0x01000000) ? 1 : 0);
-		output().set_value("stage_led_1",     (data & 0x02000000) ? 1 : 0);
-		output().set_value("stage_led_2",     (data & 0x04000000) ? 1 : 0);
-		output().set_value("stage_led_3",     (data & 0x08000000) ? 1 : 0);
-	}
-	if (ACCESSING_BITS_16_23)
-	{
-		output().set_value("stage_led_4",     (data & 0x00010000) ? 1 : 0);
-		output().set_value("stage_led_5",     (data & 0x00020000) ? 1 : 0);
-		output().set_value("stage_led_6",     (data & 0x00040000) ? 1 : 0);
-		output().set_value("stage_led_7",     (data & 0x00080000) ? 1 : 0);
-	}
+	device.execute().set_input_line(INPUT_LINE_IRQ0, ASSERT_LINE);
 }
 
-void firebeat_state::lamp_output2_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+WRITE_LINE_MEMBER(firebeat_state::ata_interrupt)
 {
-//  printf("lamp_output2_w: %08X, %08X, %08X\n", data, offset, mem_mask);
+	m_maincpu->set_input_line(INPUT_LINE_IRQ4, state);
 }
 
-void firebeat_state::lamp_output2_ppp_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+WRITE_LINE_MEMBER(firebeat_state::gcu_interrupt)
 {
-	lamp_output2_w(offset, data, mem_mask);
+	m_maincpu->set_input_line(INPUT_LINE_IRQ0, state);
+}
 
-	// ParaParaParadise lamps (active high)
-	// 0x00010000 Top LED 0
-	// 0x00020000 Top LED 1
-	// 0x00040000 Top LED 2
-	// 0x00080000 Top LED 3
-	// 0x00000001 Top LED 4
-	// 0x00000002 Top LED 5
-	// 0x00000004 Top LED 6
-	// 0x00000008 Top LED 7
-	if (ACCESSING_BITS_16_23)
-	{
-		output().set_value("top_led_0",       (data & 0x00010000) ? 1 : 0);
-		output().set_value("top_led_1",       (data & 0x00020000) ? 1 : 0);
-		output().set_value("top_led_2",       (data & 0x00040000) ? 1 : 0);
-		output().set_value("top_led_3",       (data & 0x00080000) ? 1 : 0);
-	}
-	if (ACCESSING_BITS_0_7)
-	{
-		output().set_value("top_led_4",       (data & 0x00000001) ? 1 : 0);
-		output().set_value("top_led_5",       (data & 0x00000002) ? 1 : 0);
-		output().set_value("top_led_6",       (data & 0x00000004) ? 1 : 0);
-		output().set_value("top_led_7",       (data & 0x00000008) ? 1 : 0);
-	}
+WRITE_LINE_MEMBER(firebeat_state::sound_irq_callback)
+{
 }
 
-void firebeat_state::lamp_output3_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+/*****************************************************************************/
+
+void firebeat_spu_state::machine_start()
 {
-//  printf("lamp_output3_w: %08X, %08X, %08X\n", data, offset, mem_mask);
+	m_dma_timer = timer_alloc(TIMER_SPU_DMA);
```

Comment:
I think it is preferred to avoid the common device timer callback these days, and rather route directly to your own callback function instead.